### PR TITLE
Make salesforce sync dnc on first contact download

### PR DESCRIPTION
**Pull request by galvani**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Y
| Automated tests included? | N
| BC breaks? | Y
| Deprecations? | N
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
We need to synchronize DNC record for new Lead/Contacts downloaded from SalesForce. Currently, the plugin relied solely on DNC history, which is not synchronized unless we push something and nor it is synchronized if the contact has been created *opted-out*.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:

1. Setting up SF plugin
2. Selecting “Use latest updated Do Not Contact record” under “Features” in Salesforce plugin
3. Pulling contacts from SF who are labeled as DNC out of the box into Mautic.
4. Contacts who were not already known contacts in Mautic and are labeled as DNC out of the box in SF will not show up a DNC in Mautic.

#### Steps to test this PR:
1. Load up [this PR](https://github.com/mautic-inc/mautic/tree/salesforce-dnc-new-contacts)
2. Create SF contact opted-out
3. Sync
4. Verify downloaded is DNCed

### CODE REVIEW

**Comment by galvani** [comment]
FYI: this has been reviewed by mautic inc. so testing might be the only thing needed.

**Comment by npracht** [comment]
Hi @galvani is that still `WIP`? Also, can you check travis issue?

**Comment by galvani** [comment]
@npracht no, you could have merged it, I believe recent merge broke it. I will check on it now.

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 versions (except for major security purpose) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.0.1 candidate.**

## How to do?
1. Check if the bug still present in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please comment on your PR: “This PR only applies to Mautic 2.x”

**Please report results by commenting on your PR to make us administration easier.**

_In case your bugfix only apply to Mautic 2, we'll consider adding it in an extra Mautic 2 version._
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by npracht** [comment]
Hello @alanhartless could you please test mautic 3.0.0-beta2 and tell me if the issue is still existing?
- If no, i'll tag this issue at `M2 only relevant`
- If yes, could you please rebase your branch against 3.x branch ?

Waiting for your feedback, i would very love to merge it in 3.0.1 if still relevant. Thanks !

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by npracht** [comment]
@mautic/acquia-po can someone fix this asap ? Otherwise i'll have to postpone from next release again. Thanks.

**Comment by dennisameling** [comment]
@anton-vlasenko any chance you could look into the failing tests? Thanks!

**Comment by anton-vlasenko** [comment]
@dennisameling I'm looking into them

**Comment by dennisameling** [comment]
@anton-vlasenko Though the PHPUnit tests are passing, there are some PHP warnings in there. Would you mind looking into those? Thanks 🚀 

```bash
PHP Warning - array_search() expects parameter 2 to be array, null given - in file /home/travis/build/mautic/mautic/plugins/MauticCrmBundle/Api/SalesforceApi.php - at line 114
PHP Warning:  array_search() expects parameter 2 to be array, null given in /home/travis/build/mautic/mautic/plugins/MauticCrmBundle/Api/SalesforceApi.php on line 114
PHP Warning - array_search() expects parameter 2 to be array, null given - in file /home/travis/build/mautic/mautic/plugins/MauticCrmBundle/Api/SalesforceApi.php - at line 127
PHP Warning:  array_search() expects parameter 2 to be array, null given in /home/travis/build/mautic/mautic/plugins/MauticCrmBundle/Api/SalesforceApi.php on line 127
```

and

```bash
There was 1 risky test:
1) MauticPlugin\MauticCrmBundle\Tests\Api\SalesforceApiTest::testRequestQueryAllAndHandle
This test did not perform any assertions
/home/travis/build/mautic/mautic/plugins/MauticCrmBundle/Tests/Api/SalesforceApiTest.php:360
```

We're almost there 😊

**Comment by anton-vlasenko** [comment]
@dennisameling I've fixed the errors you mentioned. Let's see if it helps.

**Comment by anton-vlasenko** [comment]
I don't have time to write additional tests right now, but I will be able to do it in 2-3 hours.

**Comment by anton-vlasenko** [comment]
I wrote 1 unit test but I don't have the capacity to write more unit tests now.
The PR is still failing. I may add more tests later.

**Comment by RCheesley** [comment]
@anton-vlasenko also a conflict to resolve so we will probably have to push this back again

**Comment by RCheesley** [comment]
This one needs to be rebased please @mautic/acquia-po

File: plugins/MauticCrmBundle/Api/SalesforceApi.php [hunk 1]
```diff
@@ -337,17 +346,7 @@ public function getLeads($query, $object)
             $fields[] = 'Id';
             $fields   = implode(', ', array_unique($fields));
```
Comment (kuzmany) [inline plugins/MauticCrmBundle/Api/SalesforceApi.php:347]: $fields are passed to to requestQueryAllAndHandle method and should be array

This return error

`Type error: Argument 2 passed to MauticPlugin\MauticCrmBundle\Api\SalesforceApi::requestQueryAllAndHandle() must be of the type array, string given`

**Review by kuzmany — CHANGES_REQUESTED** [review]

**Comment by kuzmany** [comment]
@galvani need resolve conflicts

**Comment by escopecz** [comment]
Also rebase to 5.x branch

**Comment by escopecz** [comment]
Closing in favor of https://github.com/mautic/mautic/pull/12622

**Pull request closed without merging**